### DISCUSSION TITLE
How do we delete a permission?

### DISCUSSION THREAD

**Discussion by euoia** [spatie/laravel-permission#2278] [Q&A] [answered]
Hey - thanks for creating this excellent and much used library.

What is the correct way to permissions? I have implemented this method, which seems to work:

```
    public function deletePermission($name)
    {
        $permission = Permission::findByName($name);
        if ($permission !== null) {
            $permission->delete();
            $this->info("Permission deleted: {$permission->name}");
            $this->hasChanges = true;
        }
    }
```

But my static analyser (`larastan`) is complaining that `delete` is an undefined method. And indeed if we look at `findByName` it returns a `PermissionContract` which doesn't have the `delete` method:
https://github.com/spatie/laravel-permission/blob/main/src/Models/Permission.php#L86
https://github.com/spatie/laravel-permission/blob/main/src/Contracts/Permission.php

I would think that given that `Permission` extends `Model`, that `PermissionContract` ought to extend the Model contract, or `findByName` could return a `Permission` object directly?

Thanks again!

**Comment by parallels999** [comment, marked as answer]
> that PermissionContract ought to extent the Model or findByName could return a Permission object directly?

No, beacuse model can be customizable

you may add a comment:
`/** @var YourPermisionClass $permission */`

**Reply by euoia** [reply]
Ah, so the intention is that a consumer of this library may implement `PermissionContract` in a different way? I think I understand, but I would question how useful that is.

I wonder whether a `delete` method could be added to the contract?

Adding `/** @var Permission $permission */` above `$permission = Permission::findByName($name);` does make `phpstan` happy, thanks.

**Reply by parallels999** [reply]
> does make phpstan happy

Mark this as answered
